src/state/queries/post-feed.ts

```diff
@@ -0,0 +1,171 @@
+import {AppBskyFeedDefs, AppBskyFeedPost, moderatePost} from '@atproto/api'
+import {useInfiniteQuery, InfiniteData, QueryKey} from '@tanstack/react-query'
+import {useSession} from '../session'
+import {useFeedTuners} from '../preferences/feed-tuners'
+import {FeedTuner, NoopFeedTuner} from 'lib/api/feed-manip'
+import {FeedAPI} from 'lib/api/feed/types'
+import {FollowingFeedAPI} from 'lib/api/feed/following'
+import {AuthorFeedAPI} from 'lib/api/feed/author'
+import {LikesFeedAPI} from 'lib/api/feed/likes'
+import {CustomFeedAPI} from 'lib/api/feed/custom'
+import {ListFeedAPI} from 'lib/api/feed/list'
+import {MergeFeedAPI} from 'lib/api/feed/merge'
+import {useStores} from '../models/root-store'
+
+type ActorDid = string
+type AuthorFilter =
+  | 'posts_with_replies'
+  | 'posts_no_replies'
+  | 'posts_with_media'
+type FeedUri = string
+type ListUri = string
+export type FeedDescriptor =
+  | 'home'
+  | 'following'
+  | `author|${ActorDid}|${AuthorFilter}`
+  | `feedgen|${FeedUri}`
+  | `likes|${ActorDid}`
+  | `list|${ListUri}`
+
+type RQPageParam = string | undefined
+
+export function RQKEY(feedDesc: FeedDescriptor) {
+  return ['post-feed', feedDesc]
+}
+
+export interface FeedPostSliceItem {
+  _reactKey: string
+  uri: string
+  post: AppBskyFeedDefs.PostView
+  record: AppBskyFeedPost.Record
+  reason?: AppBskyFeedDefs.ReasonRepost
+}
+
+export interface FeedPostSlice {
+  _reactKey: string
+  rootUri: string
+  isThread: boolean
+  source: undefined // TODO
+  items: FeedPostSliceItem[]
+}
+
+export interface FeedPage {
+  cursor: string | undefined
+  slices: FeedPostSlice[]
+}
+
+export function usePostFeedQuery(
+  feedDesc: FeedDescriptor,
+  opts?: {enabled?: boolean; disableTuner?: boolean},
+) {
+  const {agent} = useSession()
+  const feedTuners = useFeedTuners(feedDesc)
+  const store = useStores()
+  const enabled = opts?.enabled !== false
+
+  const api: FeedAPI = useMemo(() => {
+    if (feedDesc === 'home') {
+      return new MergeFeedAPI(agent)
+    } else if (feedDesc === 'following') {
+      return new FollowingFeedAPI(agent)
+    } else if (feedDesc.startsWith('author')) {
+      const [_, actor, filter] = feedDesc.split('|')
+      return new AuthorFeedAPI(agent, {actor, filter})
+    } else if (feedDesc.startsWith('likes')) {
+      const [_, actor] = feedDesc.split('|')
+      return new LikesFeedAPI(agent, {actor})
+    } else if (feedDesc.startsWith('feedgen')) {
+      const [_, feed] = feedDesc.split('|')
+      return new CustomFeedAPI(agent, {feed})
+    } else if (feedDesc.startsWith('list')) {
+      const [_, list] = feedDesc.split('|')
+      return new ListFeedAPI(agent, {list})
+    } else {
+      // shouldnt happen
+      return new FollowingFeedAPI(agent)
+    }
+  }, [feedDesc, agent])
+  const tuner = useMemo(
+    () => (opts?.disableTuner ? new NoopFeedTuner() : new FeedTuner()),
+    [opts?.disableTuner],
+  )
+
+  const pollLatest = useCallback(async () => {
```

Comment:
I wonder if we shouldn't just do polling, `hasNew` state, etc in this hook and expose as state 🤔  Would remove some logic from `Feed` and make that simpler.

In general, I think we should aim for something as simple as:
- fetch data and handle refresh and state up top (like in this hook)
  - all logic lives here
- construct item skeleton
- pass to FlatList to render

Maybe can't get to that point, or not right now anyway, but could be a good guiding light moving forward.

Reply:
That's probably possible, though we do need the `onHasNew` emitter so that we can emit that state back out to the wrapper components

Reply:
If outer components need state, could we move this hook up a level and flatten children out more? Then we don't need to emit anything, we just check for `hasNew` and render stuff.

But yeah not saying we should do all this here 👍

Reply:
I'd like to keep the hook inside the `<Feed>` component but maybe the answer is to move that state and UI behaviors into the Feed component